cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -54,76 +169,274 @@ theorem not_contains_of_empty {α β} [BEq α] (k : α) :
+  simp only [List.subset_def] at h₂
   exact h₂ h₁
 
+/--
+  Very similar to `mem_kvs_make` above
+-/
+theorem mem_values_make [LT α] [StrictLT α] [DecidableLT α] {xs : List (α × β)} :
+  v ∈ (Map.make xs).values → v ∈ xs.map Prod.snd
+:= by
+  -- despite the similarity to `mem_kvs_make`, the proof does not currently
+  -- use `mem_kvs_make`
+  simp only [values, make]
+  simp only [List.mem_map, forall_exists_index, and_imp]
+  intro (k, v) h₁ h₂
+  exists (k, v)
+  subst h₂
+  simp only [and_true]
+  have h₂ := List.canonicalize_subseteq Prod.fst xs
+  simp only [List.subset_def] at h₂
+  exact @h₂ (k, v) h₁
+
 theorem make_nil_is_empty {α β} [LT α] [DecidableLT α] :
   (Map.make [] : Map α β) = Map.empty
 := by simp [make, empty, List.canonicalize_nil]
 
+/--
+  Note that the converse of this is not true:
+  counterexample `xs = [(1, false)]`, `ys = []`, `ab = (1, false)`.
+-/
+theorem make_cons [LT α] [DecidableLT α] {xs ys : List (α × β)} {ab : α × β} :
+  make xs = make ys → make (ab :: xs) = make (ab :: ys)
+:= by
+  simp only [make, mk.injEq]
+  apply List.canonicalize_cons
+
 /-! ### find? and mapOnValues -/
 
-theorem find?_mapOnValues {α β γ} [LT α] [DecidableLT α] [DecidableEq α] (f : β → γ) (m : Map α β) (k : α)  :
+theorem mapOnValues_wf [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} :
+  m.WellFormed ↔ (m.mapOnValues f).WellFormed
+:= by
+  unfold mapOnValues WellFormed toList
+  sorry
+
+theorem mapOnValues_empty {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} :
+  (empty : Map α β).mapOnValues f = empty
+:= by
+  simp [mapOnValues, empty]
+
+theorem find?_mapOnValues {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α}  :
   (m.find? k).map f = (m.mapOnValues f).find? k
 := by
-  simp [Map.find?, Map.mapOnValues, Map.kvs, ←List.find?_pair_map]
-  cases __ : List.find? (fun x => x.fst == k) m.1 <;>
-  simp only [Option.map_none', Option.map_some']
+  simp only [find?, kvs, mapOnValues, ← List.find?_pair_map]
+  cases m.1.find? (λ x => x.fst == k) <;> simp
 
-theorem find?_mapOnValues_some {α β γ} [LT α] [DecidableLT α] [DecidableEq α] (f : β → γ) {m : Map α β} {k : α} {v : β} :
+theorem find?_mapOnValues_some {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α} {v : β} :
   m.find? k = .some v →
   (m.mapOnValues f).find? k = .some (f v)
 := by
   intro h₁
-  rw [← find?_mapOnValues f m k]
+  rw [← find?_mapOnValues]
   simp [Option.map, h₁]
 
-theorem find?_mapOnValues_none {α β γ} [LT α] [DecidableLT α] [DecidableEq α] (f : β → γ) {m : Map α β} {k : α} :
+theorem find?_mapOnValues_none {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α} :
   m.find? k = .none →
   (m.mapOnValues f).find? k = .none
 := by
   intro h₁
-  rw [← find?_mapOnValues f m k]
+  rw [← find?_mapOnValues]
   simp [Option.map, h₁]
 
-theorem mapOnValues_eq_make_map {α β γ} [LT α] [StrictLT α] [DecidableLT α] (f : β → γ) {m : Map α β}
```

Comment:
This is a breaking change.

Reply:
Will revert, but I'd like to understand better how this is a breaking change.

I think the only case this would be breaking is if some usage site had `mapOnValues_eq_make_map (h1 := ...)` instead of just `mapOnValues_eq_make_map h1`?  The latter is much more concise and is not broken by this change.  Am I missing something?

Reply:
Putting this in a separate PR and reverting it here